huaweicloud/config/auth.go

```diff
@@ -0,0 +1,374 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/logging"
+	"github.com/jmespath/go-jmespath"
+	"github.com/mitchellh/go-homedir"
+)
+
+const (
+	securityKeyURL     string = "http://169.254.169.254/openstack/latest/securitykey"
+	keyExpiresDuration int64  = 600
+)
+
+// CLI Shared Config
+type SharedConfig struct {
+	Current  string    `json:"current"`
+	Profiles []Profile `json:"profiles"`
+}
+
+type Profile struct {
+	Name             string `json:"name"`
+	Mode             string `json:"mode"`
+	AccessKeyId      string `json:"accessKeyId"`
+	SecretAccessKey  string `json:"secretAccessKey"`
+	SecurityToken    string `json:"securityToken"`
+	Region           string `json:"region"`
+	ProjectId        string `json:"projectId"`
+	DomainId         string `json:"domainId"`
+	AgencyDomainId   string `json:"agencyDomainId"`
+	AgencyDomainName string `json:"agencyDomainName"`
+	AgencyName       string `json:"agencyName"`
+}
+
+func buildClient(c *Config) error {
+	if c.Token != "" {
+		return buildClientByToken(c)
+	} else if c.AccessKey != "" && c.SecretKey != "" {
+		return buildClientByAKSK(c)
+	} else if c.Password != "" && (c.Username != "" || c.UserID != "") {
+		return buildClientByPassword(c)
+	} else if c.SharedConfigFile != "" {
+		return buildClientByConfig(c)
+	}
+
+	return buildClientByMeta(c)
+}
+
+func genClient(c *Config, ao golangsdk.AuthOptionsProvider) (*golangsdk.ProviderClient, error) {
+	client, err := huaweisdk.NewClient(ao.GetIdentityEndpoint())
+	if err != nil {
+		return nil, err
+	}
+
+	// Set UserAgent
+	client.UserAgent.Prepend("terraform-provider-iac")
+
+	config, err := generateTLSConfig(c)
+	if err != nil {
+		return nil, err
+	}
+	transport := &http.Transport{
+		Proxy:           http.ProxyFromEnvironment,
+		TLSClientConfig: config,
+	}
+
+	client.HTTPClient = http.Client{
+		Transport: &LogRoundTripper{
+			Rt:         transport,
+			OsDebug:    logging.IsDebugOrHigher(),
+			MaxRetries: c.MaxRetries,
+		},
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			if client.AKSKAuthOptions.AccessKey != "" {
+				golangsdk.ReSign(req, golangsdk.SignOptions{
+					AccessKey:  client.AKSKAuthOptions.AccessKey,
+					SecretKey:  client.AKSKAuthOptions.SecretKey,
+					RegionName: client.AKSKAuthOptions.Region,
+				})
+			}
+			return nil
+		},
+	}
+
+	if c.MaxRetries > 0 {
+		client.MaxBackoffRetries = uint(c.MaxRetries)
+		client.RetryBackoffFunc = retryBackoffFunc
+	}
+
+	// Validate authentication normally.
+	err = huaweisdk.Authenticate(client, ao)
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+func genClients(c *Config, pao, dao golangsdk.AuthOptionsProvider) error {
+	client, err := genClient(c, pao)
+	if err != nil {
+		return err
+	}
+	c.HwClient = client
+
+	client, err = genClient(c, dao)
+	if err == nil {
+		c.DomainClient = client
+	}
+	return err
+}
+
+func buildClientByToken(c *Config) error {
+	var pao, dao golangsdk.AuthOptions
```

Comment:
suggest replace 'pao' and 'dao' with more meaningful names

Reply:
done